### ISSUE TITLE
Integrate qualification times into the registration process

### ISSUE THREAD

**Issue by jgouly** [thewca/worldcubeassociation.org#2051]
You can currently register in events for a competition that has pre-qualification limits.
There could be some information displayed during registration to show that there is qualification per event. So you know to either not sign up, or you need to beat the qualification in an earlier competition.

**Comment by Mollerz** [comment]
I don't think it should fully prevent you from registering for the event, you might have a competition between registering and the time of the competition that allows you to qualify for it. Perhaps give organisers an option to press a button that removes all those who haven't got results with the designated qualifying times? Maybe with a check...

**Comment by jgouly** [comment]
Yup, that's why I was suggesting to show the qualification information!

**Comment by cubewhiz** [comment]
It's not a "must have," but something like this would be very helpful for big competitions like Nats or Worlds, so that we don't have to manually edit hundreds of registrations.

**Comment by cubewhiz** [comment]
This year we were able to edit the registrations for Nationals through SQL query, but many competitors (and their parents) were quite angry because they were removed from the events (and they didn't quite understand the requirements). I think it is much better to prevent them from registering for these events in the first place than it is to remove them from the events later. We would be grateful if this could be implemented by the end of 2018 so it is available for Nats 2019.

**Comment by Mollerz** [comment]
I think preventing people from registering for not making the qualification is a bad idea. In many cases these qualifications are for major championships which are announced quite far in advanced. It is very likely that someone will compete in another local competition in the mean time and possibly get some of the qualifications. In which case they will need to email the organisers and ask to be added to events.

I think if you select events with qualification in the registration procedure it should warn you if there are qualification times that you have not reached. Perhaps under the registration requirements it puts a red box that shows all qualification times, and if you are logged in it only shows the ones you have not made yet.

**Comment by ToastyKen** [comment]
@Mollerz Fyi, we've actually used the system we're recommending, and it worked better. We're not just posing a hypothetical.

For CubingUSA Nationals 2017, we rolled our own registration system, and it worked like this:

1) You could not register for events you didn't qualify in.

2) You needed to qualify in at least one event before registering at all.

3) If you were already registered, and you later qualified for an event, you could just log back in and add it (and pay the additional event fee). No email to organizers needed.

The only issue was that it took a few days for us to pick up WCA results, so people had to wait a few days to add events they just qualified for. This led to a little confusion but no significant problems, as they were able to register in a few days.

Overall, having actually used both systems, last year's system was much better because there were no nasty surprises. **Disclaimers and warnings don't work, because people don't read.** We had people show up in person at the venue this year complaining about their events being dropped despite warnings on both the WCA site and our event site about qualifying times.

**Comment by Mollerz** [comment]
Ok I agree that is a better system, in which case we do need the option on the WCA website to charge different fees based on the events offered. And we also need to be able to give the option to allow people to therefore amend their own registrations. Maybe a checkbox to allow the organisers to either allow or deny the option for competitors to change their own registrations?

**Comment by Jambrose777** [comment]
There are a couple of points to go along with this:
- Displaying qualification times on the events tab would give the competitor a consistent place to be aware of these restrictions.
- Some sort of restriction or process once the deadline is up can help WRT. We have to run custom scripts, which take time to write and ensure that the correct registrations are being removed.
- Emailing competitors notifying them that they were removed from events would help with competitor awareness.

**Comment by ToastyKen** [comment]
One note: Upon reaching the qualification deadline, we need to make sure to take into account any WCA events from before the deadline that haven't been officially submitted yet. This should be feasible since those competitions themselves should already be known.


Btw, just adding that, again this year, we had competitors who didn't read and were confused about being removed from events, including some who didn't realize what they qualified for until they showed up at the event. (These tend to be the younger competitors, where neither they nor their parents read the site carefully enough.)

**Comment by timreyn** [comment]
We'd like to have this for future large CubingUSA competitions, so I'm working on implementing this.

One note: I'm leaving a comment on the WCIF spec (https://docs.google.com/document/d/1lE70gW2rA2J2WlXeSl6N2mLnJ52ibTP_VsIwrAVdl2I/edit#) about the "when" parameter.  It's currently a DateTime, I think it would be less confusing for it to be a date (YYYY-MM-DD).  We can discuss on the comment thread over there, and I'll implement accordingly!

I'm planning to split this across a few PRs:

1. Allow setting a qualification time per event, and display it on the events list.
2. Prevent competitors from registering for events they're not qualified for.
3. Add an option for whether competitors should be allowed to edit events after their registration is accepted.

**Comment by timreyn** [comment]
Status: you can now set qualification times in the Events page. #7016 fixes a few usability bugs from that PR.

#7019 (just created) adds an option to prevent competitors from registering for events they're not qualified for. This does not apply to Top N qualifications (see below).

I have one more PR planned (not started yet) to add a button to the registration page to remove non-qualified competitors from events. This can be used if the above option is not selected, and it can also be used for Top N events. The button will not allow you to remove people before the qualification deadlines are met, and will warn if there are any pre-deadline competitions with those events with results that haven't been posted. It will notify all affected competitors by email.

Why not use this for Top N? Let's take top 16 for 333fm. You could imagine a system where, if 16 people are registered for 333fm and someone new registers who is better than one of them, the person at the bottom gets knocked out when that person registers. But this has two downsides:
- People need to remember to register when they break a PR, and they might need to register multiple times after getting knocked out multiple times.
- What happens if someone drops out a day before the registration deadline? Then a spot opens up, not to the 17th best person, but to whoever notices that there's an open spot in time to register.
So I think it's easier to just let anyone with a result register, and then after the qualification deadline, remove all but the top N people.

**Comment by sarahstrong314** [comment]
Amazing! This will save A LOT of time and resources for championship competitions. Thanks, Tim!

The process of determining who qualified for which event was tedious. It wasn't as simple as looking at the psych sheet at the end of the registration period since competitors may have competed in competitions that took place after the qualification deadline but before registration ended, and there's also the possibility that competitions that took place before the qualification deadline haven't had their results posted yet (albeit this is very unlikely). At least one of the organizers needed knowledge of how to use the WCA DB export unless they wanted to check by hand for these edge cases.

Removing events from registrations one by one is also a tedious process. For NA Champs 2022, we manually removed ~900 events from ~300 competitors, which took around 90 minutes. There wasn't a quick way to double check either. For NA Champs 2022 I did a diff comparison of the registration CSV before and after event removals and checked the results against the program I used to determine unqualified results.

Another time, resource, and money saver from these new changes will be preventing competitors who don't have a WCA ID from registering and paying. For NA Champs 2022 we had ~90 registrations from people who didn't compete before the qualification deadline and most of them paid the registration fee. Reimbursing their registrations cost us a considerable amount in Stripe fees. We also had to deal with a lot of emails from these registrants who were unclear about how qualification times worked (e.g. many registrants thought they could qualify without competing since they already paid, and others thought that if you qualified or one event you qualified for all of them). Not allowing registrants to register for events they're not qualified in will make this much more clear for them and allow them to make travel plans accordingly.

**Comment by dunkOnIT** [comment]
Qualifications are now implemented
